zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -383,16 +383,59 @@ pub(crate) fn get_transparent_balances<P: consensus::Parameters>(
 pub(crate) fn add_transparent_account_balances(
     conn: &rusqlite::Connection,
     mempool_height: BlockHeight,
+    min_confirmations: u32,
```

Comment:
In the past, we have specifically chosen to use a minconf of 0 for transparent balances, because the mobile wallets don't permit spending transparent funds except to shield them. Given that the funds aren't available to the wallet until they've been shielded, we allow shielding at minconf 0 because it's the _shielding_ transaction whose number of confirmations we're concerned with.

Reply:
Thanks for the context. I'm not clear what is the recommendation here. Let me put in words what I understand. from your comment. 

I assume that you are pointing out that, at the moment, `min_confirmations` is always **zero** for transparent funds. Given transparent funds are currently being shielded before they can be spent and shielding transactions use zero confirmations for the t -> 🛡️ part and non-zero `min_confirmations` for spending shielded funds. This would mean that the parameter `min_confirmations` for this particular functions is not correct in the current context and assumptions.

Reply:
Yes, `min_confirmations` is not being used for transparent TXOs, and that's a bug that this PR fixes. But @nuttycom is incorrect in asserting that "we allow shielding at minconf 0 [in the mobile wallets]". We do not, for two reasons:

1. `minimumShieldingConfirmations` is set to 1 in the SDKs, not 0.
2. None of the balances include unmined UTXOs. That also applies to `unshielded_balance` before this PR. The reason is this WHERE clause condition (line 395 before / line 397 after):
```sql
            t.mined_height < :mempool_height -- tx is mined
```
$\hspace{1.6em}$ And so we do not see unmined UTXOs as available to shield.

If we want to change this then we should do so separately.

Reply:
I think I may have addressed this as the test `transparent_balance_spendability` passes. I still need to make the test code non-repetitive (extract the `check_balance` function).

I enforced very obvious to the eye checks for zero and one confirmations with these kind of checks

```
    check_balance(&st, 0, &Balance::ZERO);
    // we currently treat min_confirmation the same regardless they are 0 or 1
    check_balance(&st, 1, &Balance::ZERO);
 ```
 
 and also verified that the balances hold when it comes to moving the min_confirmations up by one or mining an extra block.

Reply:
It wasn't correct to conclude that `transparent_balance_spendability` passes, because the wrong test was being run. We'll see once CI has finished.

Reply:
It doesn't in fact pass:
```
---- wallet::transparent::tests::transparent_balance_spendability stdout ----
thread 'wallet::transparent::tests::transparent_balance_spendability' panicked at /home/runner/work/librustzcash/librustzcash/zcash_client_backend/src/data_api/testing/transparent.rs:56:5:
assertion `left == right` failed
  left: Some(Zatoshis(100000))
 right: Some(Zatoshis(0))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    wallet::transparent::tests::transparent_balance_spendability

test result: FAILED. 98 passed; 1 failed; 1 ignored; 0 measured; 0 filtered out; finished in 41.27s
```

Reply:
Filed #1602 for the missing traceback.

Reply:
```
---- wallet::transparent::tests::transparent_balance_spendability stdout ----
thread 'wallet::transparent::tests::transparent_balance_spendability' panicked at /media/ssd/home/daira/research/librustzcash/zcash_client_backend/src/data_api/testing/transparent.rs:56:5:
assertion `left == right` failed
  left: Some(Zatoshis(100000))
 right: Some(Zatoshis(0))
stack backtrace:
   0: rust_begin_unwind
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/std/src/panicking.rs:647:5
   1: core::panicking::panic_fmt
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:342:17
   3: core::panicking::assert_failed
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/panicking.rs:297:5
   4: zcash_client_backend::data_api::testing::transparent::check_balance
   5: zcash_client_backend::data_api::testing::transparent::transparent_balance_spendability
             at /media/ssd/home/daira/research/librustzcash/zcash_client_backend/src/data_api/testing/transparent.rs:361:5
   6: zcash_client_sqlite::wallet::transparent::tests::transparent_balance_spendability
             at ./src/wallet/transparent.rs:890:9
   7: zcash_client_sqlite::wallet::transparent::tests::transparent_balance_spendability::{{closure}}
             at ./src/wallet/transparent.rs:889:42
   8: core::ops::function::FnOnce::call_once
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/ops/function.rs:250:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/aedd173a2c086e558c2b66d3743b344f977621a7/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Reply:
There were a couple of things here:
1. check_balance was not properly using `min_confirmations`. Anyway the test still failed. then,
2. the `transparent_balance_spendability` test was not advancing and scanning the chain so the balance was being reported properly by the backend (from it's PoV) and it was the test that assumed otherwise. I added the blocks and scanned them and the test passed.